### PR TITLE
Introduce jailed behavior for admin area

### PR DESCRIPTION
**Pull request by dgrammatiko**
##### Why?

When editing an article, menu, category (anything singular), and press the back button, the refresh or the close button on the browser, user most probably will loose any data without warning. This PR throws a standard browser window with custom message to inform him/her. Of course we are not forcing anything, so he/she can go ahead and close the window, but at least we did informed them!
There is a check for debug and language debug, so this functionality is enabled only when none of these is enabled!
##### Preview:

Safari
screen shot 2015-01-19 at 1 37 39

Chrome
screen shot 2015-01-19 at 1 37 16

Firefox (NOTE: FF doesn’t support custom messages in the alert called onbeforeunload!)
screen shot 2015-01-19 at 1 38 23

Opera
screen shot 2015-01-19 at 1 37 59

And IE 8 (YES it even works on IE

### CODE REVIEW

File: libraries/cms/html/behavior.php [hunk 1]
```diff
@@ -92,6 +92,56 @@ public static function core()
...
+		}
+
+		// Include jQuery
+		JHtml::_('jquery.framework');
+
+		JFactory::getDocument()->addScriptDeclaration("
+		jQuery(document).ready(function() {
+			// Prevent alert for links
+			jQuery('a').bind('mousedown', function () {
+				jQuery(window).off('beforeunload');
```
Comment (Fedik) [inline libraries/cms/html/behavior.php]: and what if another extension use `beforeunload` event? :smile:
try use namespaced events, like `beforeunload.jailed`

**Comment by Fedik** [comment]
you approach a bit confuse, 
the function will display alert even nothing edited, can be annoying.
what about something like:

```javascript
jQuery(document).ready(function() {
  var changed;
  jQuery('[name]').on('change input', function(){
    changed = true;
  });
  jQuery(window).on('beforeunload', function (event) {
   if(changed){
      return 'alert';
    }
  });
});
```
or around :smile:

File: libraries/cms/html/behavior.php (hunk 1 above)
Comment (dgt41) [inline libraries/cms/html/behavior.php]: Agree!

**Comment by brianteeman** [comment]
I really appreciate the detailed work that has gone into this and for testing on multiple browsers. However personally I dont like this. IF it was only happening when some data was changed I might think differently but for it to happen all the time is just a pain

**Comment by dgt41** [comment]
@Fedir @brianteeman Let’s say you try to edit an article. You click on the title in the list view and once you have the form you realize that this is NOT the one you want and you click the back button. Without throwing an alert (actually joomla is politely asking you to use save or cancel button) you will end up with a lock icon in the article, which means that you have to checkin that article (if you are allowed to do so!).
I hope this explains that this is not only for end users but also a way to keep Joomla admins out of the check in button!

**Comment by brianteeman** [comment]
You still get the lock icon with this popup if you click leave this page

**Comment by dgt41** [comment]
@brianteeman Sadly thats true

**Comment by dgt41** [comment]
@Fedik @brianteeman Implemented the check for changes! So the alert will NOT trigger if no changes made!

**Comment by brianteeman** [comment]
Just retested on a clean site and it didnt matter what I did I never got the popup

File: libraries/cms/html/behavior.php [hunk 2]
```diff
@@ -92,6 +92,63 @@ public static function core()
...
+			});
+
+			// Prevent alert for forms
+			jQuery('input[type=submit]').bind('mousedown', function () {
+				jQuery(window).off('beforeunload.jailed');
+			});
+
+			// Prevent alert for unchanged forms
+			jQuery('form :input').change(function() {
+				jQuery(this).closest('form').data('changed', true);
```
Comment (Fedik) [inline libraries/cms/html/behavior.php]: try replace to `jQuery(this).parents('form').attr('data-jailed', 'changed');`

File: libraries/cms/html/behavior.php [hunk 3]
```diff
@@ -92,6 +92,63 @@ public static function core()
...
+			});
+			// Alert on unintentional exit
+			jQuery(window).on('beforeunload.jailed', function (event) {
+				if(jQuery(this).closest('form').data('changed')) {
```
Comment (Fedik) [inline libraries/cms/html/behavior.php]: and then here `if(jQuery('form[data-jailed=\"changed\"]').length) {`

**Comment by Fedik** [comment]
**test**
works good for me, now

but question about whether we need it, 
maybe need some vote? :smile:

**Comment by brianteeman** [comment]
I still cant get it to work (but even if I did I would not be in favour of
this)

On 19 January 2015 at 13:38, Fedir Zinchuk <notifications@github.com> wrote:

> *test*
> works good for me, now
>
> but question about whether we need it,
> maybe need some vote? [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/5815#issuecomment-70493115>.
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by dgt41** [comment]
@brianteeman Due to the way .change is defined in jQuery, this might not work in cases where the edited element hasn’t lost its focus (you have to click somewhere else in the page)

> The change event is sent to an element when its value changes. This event is limited to <input> elements, <textarea> boxes and <select> elements. For select boxes, checkboxes, and radio buttons, the event is fired immediately when the user makes a selection with the mouse, but for the other element types the event is deferred until the element loses focus.
From jQuery [API](http://api.jquery.com/change/)

**Comment by Fedik** [comment]
@dgt41 another suggestion use `jQuery('form :input').on('changed input', function() {` instead of `.changed()`
this will mark changes on input modification, not only when input lost focus

**Comment by dgt41** [comment]
@Fedik Disabled on debug was an old request from many PLT members, you can see the comments at #4773

**Comment by Fedik** [comment]
@dgt41 so you know why it not very welcome feature? :smile:

**Comment by dgt41** [comment]
@Fedik I think this is a welcome feature for average users. Advanced users, and devs dislike anything that gets in their way, I guess 😄

**Comment by brianteeman** [comment]
I am setting this to needs review so that the CMS maintainers and PLT can decide if this concept is worth pursuing<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5815">issues.joomla.org/joomla-cms/5815</a>.</sub>

**Comment by wilsonge** [comment]
Personally I like the concept (this isn't a formal PLT opinion ;) ). However I'd make it much simpler. Just check if someone has altered the text in the textarea. I don't think changing other elements of the form is as annoying/time consuming to replace as the article itself.

**Comment by anibalsanchez** [comment]
I think that's a wonderful idea....however it **does not work** if you change text in the editor, category select (Chosen), or Publishing Dates (Calendars).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5815">issues.joomla.org/joomla-cms/5815</a>.</sub>

**Comment by dgt41** [comment]
@anibalsanchez reading others people approach in the net, most of them serialize the form data on the page load and then compare the changes (if any). Also this will prevent another erratic behavior e.g. add a word and the delete it, right now this will count as changed form, although it isn’t. Maybe the best approach is to utilize some jquery plugin for [this](https://github.com/ianand2/Form-Savior)

**Comment by anibalsanchez** [comment]
#dgt41 From my humble experience, I have created this https://github.com/anibalsanchez/jquery.saveform.js/blob/master/jquery.xtsaveform.js, based on jquery.saveform.js. It can be tweaked to include hidden fields (.xtd-save-this class).

**Comment by dgt41** [comment]
I am going to close this one, as there is little interest on improving UX.

**Comment by JoshuaLewis** [comment]
As stated above, if it is activated with altered text I'm in favor of this.

**Pull request closed without merging**